src/t_zset.c

```diff
@@ -3555,19 +3561,23 @@ void zrangeGenericCommand(zrange_result_handler *handler, int argc_start, int st
     long opt_end = 0;
     int opt_withscores = 0;
     long opt_offset = 0;
-    long opt_limit = -1;
+    long opt_limit = -1; /* -1 means all. */
+    int limit_is_set = 0; /* Determine whether the limit variant is set. */
 
     /* Step 1: Skip the <src> <min> <max> args and parse remaining optional arguments. */
     for (int j=argc_start + 3; j < c->argc; j++) {
         int leftargs = c->argc-j-1;
         if (!store && !strcasecmp(c->argv[j]->ptr,"withscores")) {
             opt_withscores = 1;
         } else if (!strcasecmp(c->argv[j]->ptr,"limit") && leftargs >= 2) {
-            if ((getLongFromObjectOrReply(c, c->argv[j+1], &opt_offset, NULL) != C_OK) ||
-                (getLongFromObjectOrReply(c, c->argv[j+2], &opt_limit, NULL) != C_OK))
-            {
+            limit_is_set = 1;
+            if (getRangeLongFromObjectOrReply(c, c->argv[j+1], 0, LONG_MAX,
+                                              &opt_offset, "offset should be greater than or equal to 0") != C_OK)
+                return;
+
+            if (getRangeLongFromObjectOrReply(c, c->argv[j+2], -1, LONG_MAX,
+                                              &opt_limit, "limit should be greater than or equal to -1") != C_OK)
```

Comment:
```suggestion
            if (getLongLongFromObjectOrReply(c, c->argv[j+2], &opt_limit, NULL) != C_OK)
```
Allow any negative value to be "all" (backwards compatibility).